testsuite/tests/lib-reachable/testopt.ml

```diff
@@ -0,0 +1,67 @@
+(* TEST
+   * native
+*)
+
+(* todo: TEXSTXX
+
+include dynlink
+libraries = ""
+files = "plugin.ml testopt.ml"
+
+* native
+** shared-libraries
+*** setup-ocamlopt.opt-build-env
+**** ocamlc.opt
+program = "${test_build_directory}/testopt.opt"
+flags = "-linkall"
+include ocamlcommon
+libraries += "dynlink"
+all_modules = "testopt.ml"
+***** run
+******check-program-output
+*)
+
+let x = Array.make 100 1234
+let y = Array.make 200 5678
+
+module Foo : sig
+  val junk : string list
+  val junk2 : string list
+end = struct
+  let junk = ["abc"; "def" ^ "yz"]
+  let junk2 = ["abc"; "def" ^ "yz"]
+end
+
+let field_path name var =
+  let arr = Gc.field_path var in
+  Printf.printf "Field path of %s =" name;
+  Array.iter (fun item -> Printf.printf " %d" item) arr;
+  Printf.printf "\n%!"
+
+let _ = field_path "x" x
+let _ = field_path "y" y
+
+let _ = field_path "Foo.junk" Foo.junk
+let _ = field_path "Foo.junk2" Foo.junk2
+
+(*
+let () =
+  try
+    Dynlink.loadfile "plugin"
+  with
+  | Dynlink.Error error ->
+    prerr_endline (Dynlink.error_message error)
+    let x = Array.make 100 "abcd"
+*)
+
+let f () = Gc.print_reachable 100
+
+let g () =
+  let x = Array.make 1000 123 in
+  let y = Array.make 2000 456 in
+  let z = Array.make 3000 789 in
+  f ();
+  if x = y || x = z then  (* ensure fields aren't optimized away *)
+    Printf.printf ""
+
+let _ = g ()
```

Comment:
Looks like flambda may be optimizing away some of my variables.  Perhaps someone can suggest something that will prevent the optimization so it generates the same output for flambda?  The test in the first Travis job shows:

```
 Field path of x = 0
-Field path of y = 1
+Field path of y = 0
 Field path of Foo.junk = 2 0
 Field path of Foo.junk2 = 2 1
 Globals:
        101 field   0 camlTestopt__gc_roots
-       201 field   1 camlTestopt__gc_roots
+       201 field   0 ???
 
 Called from Testopt.g in file "testopt.ml", line 63, characters 2-6
       2001 field   0 (frame 0)
```

Compared to `testopt.reference`:

```
Field path of x = 0
Field path of y = 1
Field path of Foo.junk = 2 0
Field path of Foo.junk2 = 2 1
Globals:
       101 field   0 camlTestopt__gc_roots
       201 field   1 camlTestopt__gc_roots

Called from Testopt.g in file "testopt.ml", line 63, characters 2-6
      2001 field   0 (frame 0)
      1001 field   1 (frame 0)
      3001 field   2 (frame 0)
```